### ISSUE TITLE
scan/fold should not consider the `init` args for scanning over

### ISSUE THREAD

**Issue by rjpower** [stanford-crfm/haliax#92]
`haliax.scan` and `fold` take an arbitrary set of arguments, and include `NamedArray` or `jax.array` as part of the scan:

https://github.com/stanford-crfm/haliax/blob/main/src/haliax/hof.py#L59

This can be confusing if you pass an array as part of the `init` argument. This should probably be exempted from the scan by default.

**Comment by dlwh** [comment]
So I'm kinda confused because we treat init differently through the entire code path https://github.com/stanford-crfm/haliax/blob/main/src/haliax/hof.py#L99-L136

**Comment by dlwh** [comment]
This passes?

```
def test_scan_doesnt_scan_init():
    Height = Axis("Height", 10)
    named1 = hax.random.uniform(PRNGKey(0), (Height,))

    init = jnp.arange(Height.size, dtype=jnp.float32)

    def scan_fun(acc, z, x):
        out = acc + z * x, x * z
        return out

    total, selected = hax.scan(scan_fun, Height)(init, 4.0, named1)

    assert jnp.all(jnp.isclose(total, init + jnp.sum(named1.array * 4.0)))

    # double check with named array init
    total, selected = hax.scan(scan_fun, Height)(hax.named(init, "Height"), 4.0, named1)

    assert jnp.all(jnp.isclose(total.array, init + jnp.sum(named1.array * 4.0)))
```

**Comment by rjpower** [comment]
Okay, that's puzzling... I didn't pick up on that originally. I found I needed to change the `is_scanned` callback to something like:

```
 def _is_block(leaf):
            return hax.is_named_array(leaf) and leaf.axes[0] == block_axis
```

for this call:

```
hax.fold(_loss_fn, axis=block_axis, is_scanned=_is_block)(
            (initial_loss, teacher_x),
            (student_layers, teacher_layers),
            key=key,
            mask=batch.attn_mask,
        )
```

Where the layers are just the stacked trees:

```
student_layers = student.transformer.layers.stacked
teacher_layers = teacher.transformer.layers.stacked
```

Without this, I kept getting `scan` errors about inputs of size `[32, 32...., 2]`. I couldn't find a model parameter that would explain this, so I assumed it was something to do with the init tuple. When I only allowed the `Block` axis, the problem went away. You're right, I can't see anything in the `scan` logic that would glom the `init` argument in otherwise....

**Comment by dlwh** [comment]
Oh the key is going to get scanned too. You should split it to Blocks.size.

**Comment by rjpower** [comment]
Argh, that's it, I should have picked up on that quicker.
